---
title: "Calculating microbial functional diversity with funfunfun"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Calculating microbial functional diversity with funfunfun}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(funfunfun)
#devtools::load_all() # for development only
```

# Introduction

funfunfun is an R package to calculate functional diversity using [Contribution Evenness](https://www.biorxiv.org/content/10.1101/2020.04.22.054593v1), a novel method described by Royalty and Steen which is tailored for microbial ecosystems.

# Theory

A functional redundancy measurement tailored for community-aggregated parameters, or the summed ecosystem processes contributed by species. First, the relative contribution that species make towards a community-aggregated parameter is determined by taking the product of species-specific trait level and the species respective abundance. This value is normalized by the summed total from across the community to determine the relative contribution by species to a community-aggregated parameter. The effective number of species contributing to a community-aggregated paramter is  calculated by passing the relative contribution distribution to the HCDT entropy equation, where entropy is and converted into an effective richness (Jost 2006). The effective richness of species contributing to a community-aggregated parameter is then normalized by the maximum theoretical number species which could contribute to the community-aggregated parameter (i.e., the community richness). This value corresponds to the proportion of the community contributing the community-aggregated parameter. Last, this value is made unrelated to species richness, which allows for comparisons between communities with different richness (Chao & Ricotta 2019).  

# Use
For calculating functional redundancy when species-specific trait data (e.g., gene copies, chemical transformation rates, etc.) and species-abundance data is available.


## Input file

We use two input files:

* The **abundance matrix** is a table containing a matrix of species abundances. Unique species are listed in columns, and unique sites are listed in rows.
* The **trait matrix** is a table containing trait levels for the same species as are in the abundance matrix, also in rows. An arbitrary number of traits are listed in columns.

For this vignette, the abundance matrix is `species_abund` and the trait matrix is `trait_levels`.

Species relative abundance data was downloaded from https://doi.org/10.6084/m9.figshare.5188273. Relative abundances were converted into count data by multiplying site sequencing effort into metagenome-assembled genome relative abundances.

Trait data represents the gene copy numbers for different nitrogen-transforming pathways in metagenome-assembled genomes downloaded from https://doi.org/10.6084/m9.figshare.5188273.
 
```{r}
abundance.matrix <- read.csv('data/MAG_abundance_table.csv', row.names = 1 )
abundance.matrix <-
  round(species.abund / min(species.abund[species.abund > 0]))
sample.effort <- min(rowSums(abundance.matrix))
abundance.matrix <- vegan::rrarefy(abundance.matrix, sample.effort)
abundance.matrix <-
  sweep(abundance.matrix, 1, rowSums(abundance.matrix), '/')
# trait.matrix <- read.csv('data/MAG_enzyme_gene_copies.csv', row.names = 1 )
fr <- royalty_fr(abundance.matrix, trait.levels, q = 0.5)
fr <-
  tidyr::separate(fr,
                  sample,
                  into = c("site", "size_fraction", "depth"),
                  sep = '_')
ggplot2::ggplot(fr, ggplot2::aes(x = trait, y = fr, color = depth)) +
  ggplot2::geom_boxplot() +
  ggplot2::ylim(0, 0.7)  
```

